design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+    the `Secret` for the `ServiceAccount` that are required for running the
+    controller in [host aware
+    mode](https://github.com/crossplane/crossplane/blob/master/design/one-pager-host-aware-stack-manager.md).
+
+The process for a `PackageInstall` is very similar, but the packages using the
+templating controller have the additional step of first producing a
+`StackDefinition` in the install `Job`, then translating that to a `Package` in
+the [_`StackDefinition`
+controller_](https://github.com/crossplane/crossplane/blob/6fc50822fbf11a7d31f8a9dabde5c8948c3b36ac/pkg/controller/packages/templates/stackdefinition_controller.go#L53).
+
+### Future Workflow
+
+The first step towards supporting the upgrade of packages is by introducing the
+`PackageRevision` type. The `PackageRevision` type will look very similar to the
+current `Package`, but will add the following fields:
+
+- `spec.infrastructureDefinitions`: list of `InfrastructureDefinition`s
+  installed by the package.
+- `spec.compositions`: list of `Composition`s installed by the package.
+- `spec.state`: the desired state, which can be one of `Pending`, `Active`,
+  `Stopping` or `Inactive`.
+
+The following fields will be removed:
+
+- `spec.packageType`: currently is one of `Provider`, `Stack`, `Application`, or
+  `Addon`.
+- `spec.permissionType`: currently is one of `Cluster` or `Namespace`.
+
+In addition, `PackageRevision` will be a cluster-scoped type, rather than the
+current namespace-scoped `Package`. The current schema for `Package` can be
+viewed
+[here](https://doc.crds.dev/github.com/crossplane/crossplane/cluster/charts/crossplane-types/crds/packages.crossplane.io_packages.yaml@v0.12.0).
+
+In order to continue to support the current functionality of
+`ClusterPackageInstall`, `PackageInstall`, `StackDefinition` the underlying
+`Package` type will be renamed to `LegacyPackage`. The existing `Package` type
+will be refactored to look like the current `ClusterPackageInstall` schema with
+thw following modifications.
+
+This field will be added:
+
+- `status.currentHash`: a string that is the `sha256` hash of the image in the
+  current `PackageRevision`.
+- `status.currentImage`: a string that is the name of the current
+  `PackageRevision` image.
+
+The following fields will be removed:
+
+- `status.installJob`: a reference to the install `Job` for the
+  `ClusterPackageInstall`. 
+- `status.packageRecord`: a reference to the `Package` that represents the
+  installation unit for the `ClusterPackageInstall`. The `PackageRevision` is
+  now discoverable by name at the cluster scope.
+
+
+Before outlining the functionality of the new `Package` and `PackageRevision`
+controllers, it is useful to establish the responsibilities of each.
+
+The `Package` controller is responsible for:
+
+- Managing the lifecycle of its `PackageRevision`s
+- Managing the lifecycle of any objects installed by its `PackageRevision`s
```

Comment:
@negz this depends on how we want to do upgrades. I like the idea of the `Package` managing the objects because it creates a clean separation in my mind about things that persist from one revision to the next vs things that do not (i.e. the controller `Deployment`). If we shift this responsibility to the `PackageRevision` I think the management of these objects gets harder. For instance, if a new revision drops support for an alpha CRD in the previous revision, how are the `PackageRevision`s supposed to work this out? The current `PackageRevision` would presumably forfeit control of its CRDs, then the new `PackageRevision` would assume control, but you would then have the potential for some CRDs to be orphaned. In addition, if we want to build out more robust upgrade strategies in the future it seems we would have more flexibility to do so if a `Package` was always knowledgeable of both its current and pending CRDs (and composition objects).